pyth/client/src/middleware_cache.rs

```diff
@@ -1,202 +1,171 @@
+        base_url: String,
+        ids: NonEmpty<I>,
+    ) -> HashMap<PathBuf, std::vec::IntoIter<Vec<Binary>>>
     where
         I: IntoIterator + Lengthy + Clone,
         I::Item: ToString,
-        T: ToString,
     {
+        let mut stored_vaas = HashMap::new();
+
         // Load data for each id.
         for id in ids.into_inner() {
-            let filename = self.create_file_name(&id);
+            let element = id.to_string();
+            let filename = Self::cache_filename(&element);
 
+            #[allow(clippy::map_entry)]
             // If the file is not in memory, try to read from disk.
-            if !self.stored_vaas.contains_key(&filename) {
-                let cache_file = DiskPersistence::new(filename.clone().into(), true);
+            if !stored_vaas.contains_key(&filename) {
+                let cache_file = DiskPersistence::new(filename.clone(), true);
 
                 // If the file does not exists, retrieve the data from the source.
                 if !cache_file.exists() {
-                    let mut pyth_client = PythClient::new(base_url.to_string());
-
-                    // Retrieve 15 samples of the data.
-                    let mut values = vec![];
-                    let shared =
-                        pyth_client.run_streaming(NonEmpty::new(vec![id.to_string()]).unwrap());
-                    while values.len() < 15 {
-                        let vaas = shared.replace(vec![]);
-                        if !vaas.is_empty() {
-                            values.push(vaas);
+                    let rt = Runtime::new().unwrap();
+                    let values = rt.block_on(async {
+                        let client = PythClient::new(base_url.clone());
+
+                        let mut stream = client
+                            .stream(NonEmpty::new(vec![id.to_string()]).unwrap())
+                            .await
+                            .unwrap();
+
+                        let mut values = vec![];
+                        while values.len() < 15 {
+                            if let Some(vaas) = stream.next().await {
+                                values.push(vaas);
+                            }
                         }
-                        sleep(Duration::from_millis(1200));
-                    }
 
-                    pyth_client.close();
+                        values
+                    });
 
                     // Store the data in the cache.
-                    self.store_data(id, values).unwrap();
+                    Self::store_data(element, values).unwrap();
                 }
 
                 // Load the data from disk.
                 let loaded_vaas = cache_file.load::<Vec<Vec<Binary>>>().unwrap();
-                self.stored_vaas
-                    .insert(filename.clone(), loaded_vaas.into_iter());
+                stored_vaas.insert(filename, loaded_vaas.into_iter());
             }
         }
+
+        stored_vaas
```

Comment:
Good place to change the published time if we need multiple different values from Pyth.